aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu

```diff
@@ -636,6 +636,11 @@ Tensor _sparse_csr_sum_cuda(const Tensor& input, IntArrayRef dims_to_sum, bool k
     [&] {
       result = reduce_sparse_csr_cuda_template<scalar_t>(input_, dims_to_sum, keepdim, ReductionAddOp<scalar_t>());
     });
+  auto is_integral =
+      !dtype.has_value() && at::isIntegralType(dtype_, /*includeBool=*/false);
+  if (is_integral) {
+    result = result.to(ScalarType::Long);
```

Comment:
So the [dtype argument for sum](https://pytorch.org/docs/master/generated/torch.sum.html?highlight=sum#torch.sum) is supposed to be used to help prevent data type overflow. If we've already performed the operation and then cast to Long, it won't help with that. We can either add support for this to `reduce_sparse_csr_cuda_template` or explicitly error out on the feature and tell the user it's not supported.

This explicit conversion at the end is a) costly and b) not achieving what the user wants.

Reply:
Hi @cpuhrsch , thank you for the comments. Let me clarify this.

[dtype argument for sum](https://pytorch.org/docs/master/generated/torch.sum.html?highlight=sum#torch.sum) is indeed used to help prevent data type overflow. If `dtype` is used and `dtype.has_value() == True`, then the logic above will not cast to Long anymore. Only when `dtype` is not used, the default data type of return value will cast to Long if the input data type is integral.

Reply:
Oh, right. Hm, but isn't that worse? That'll cause another kernel launch. We can still accumulate in a lower precision dtype, but preallocate the required output precision and convert + write into it within the CUDA kernel.

I don't think the dense counterpart of sum does additional conversion, but I might be wrong. I think it allocates in the appropriate dtype and then has the kernel write into it.

Reply:
Yes. You're right. I have updated the code to remove `to(Long)` and update input to `Long`, so that the output would be `Long` as well.

Reply:
Hi @cpuhrsch , could you please help review this PR again? Thank you!